qmail-start.8
.TH qmail-start 8
.SH NAME
qmail-start \- turn on mail delivery
.SH SYNOPSIS
.B qmail-start
[
.I logger arg ...
]
.SH DESCRIPTION
.B qmail-start
invokes
.BR qmail-send ,
.BR qmail-lspawn ,
.BR qmail-rspawn ,
and
.BR qmail-clean .
These four daemons cooperate to deliver messages from the queue.

.B qmail-start
arranges for
.BR qmail-send 's
activity record to be sent to
.BR qmail-start 's
output.
See
.B qmail-log(5)
for the format of the activity record.
Other than this,
.B qmail-start
does not print anything, even on failure.

If
.I logger
is supplied,
.B qmail-start
invokes
.I logger
with the given arguments,
and feeds
.BR qmail-send 's
activity record through
.IR logger .

Environment variables given to
.B qmail-start
will eventually be passed on to
.BR qmail-alias ,
so make sure to clean up the environment if you run
.B qmail-start
manually:

.EX
   # env - PATH="/var/qmail/bin:$PATH"
.br
   qmail-start logger -p mail.info -t qmail &
.br
   (all on one line)
.EE

.B qmail-start
sets the uid and gid of each daemon properly.

Note that
.B qmail-send
normally juggles several simultaneous deliveries.
To reduce
.BR qmail-send 's
impact on other programs,
you can run
.B qmail-start
with a low priority.
.SH "SEE ALSO"
logger(1),
nice(1),
qmail-log(5),
qmail-alias(8),
qmail-clean(8),
qmail-lspawn(8),
qmail-rspawn(8),
qmail-send(8)


CHANGES
19970220 qmail 1.00.
19970219 change: various documentation tweaks.
19970218 change: updated THOUGHTS.
19970218 change: talked about SPAWN_NUMD in FAQ. tnx EC.
19970210 change: noted in maildir.5 that readers should skip any name
         starting with a dot. tnx SP.
19970209 change: added note to splogger.8 about reliability. tnx BT.
19970209 change: added section to FAQ on slow sendmail switch. tnx BT.
19970206 change: added section to FAQ about dtcm. tnx PJG.
19970206 change: tweaked maildir.5.
19970201 change: added MH spost note to FAQ. tnx TU.
19970131 change: reorganized FAQ.
19970131 change: added web references to FAQ.
19970124 change: tweaked qmail-upgrade man page.
19970120 qmail 0.96, gamma.
19970120 change: removed various try* in auto-configuration.
19970120 bug: qmail-inject fails to quote argument addresses. impact:
         addresses containing special characters could be misinterpreted
	 or rejected. tnx C2F. fix: use quote2().
19970120 portability problem: ESIX puts syslog() and openlog() into
         -lgen. impact: can't compile under ESIX. fix: put -lgen into
	 LIBS for unix_sv. tnx RN.
19961221 qmail 0.95, gamma.
19961218 change: added various try* to TARGETS. tnx SA.
19961216 change: clarified in qmail-send.8 that virtualdomains does not
         apply to domains listed in locals.
19961216 change: slurpclose() now closes fd on out-of-memory. makes it
         more widely applicable.
19961215 change: replaced elm instructions in INSTALL.mbox with an
         explanation of what source change to make. tnx AB.
19961212 portability problem: under NEWS-OS, time_t needs sys/types.h.
         impact: couldn't compile under NEWS-OS. fix: include
	 sys/types.h in predate.c. tnx TU.
19961211 change: used timeoutread, timeoutwrite in remoteinfo(). tnx
         REB.
19961210 portability problem: apparently some SGIs produce a systype of
         irix64. impact: couldn't compile on those systems. fix: handle
	 irix64 in make-cmds. tnx M3S.
19961208 change: added note to maildir2mbox.1 about mbox locking.
19961208 qmail 0.94, gamma.
19961207 change: added QMAILDEFAULTDOMAIN, QMAILDEFAULTHOST,
         QMAILIDHOST, QMAILPLUSDOMAIN. tnx BTW.
19961206 cleanup: readsubdir() protects itself against name overflow,
         rather than depending on caller.
19961204 change: changed FAQ 7.3 to prohibit fixup relaying.
19961203 change: added note to FAQ about possibly having to put a space
         before "$SENDER" for uux. tnx FW.
19961202 change: added FAQ entry on QUEUE_EXTRA.
19961202 change: added FAQ entry on backups. tnx DP.
19961202 change: added note to INSTALL.mbox about qpopper. tnx VV.
19961201 change: replaced logger with splogger in FAQ 5.1. tnx FPL.
19961201 change: used netmask example for tcpcontrol in FAQ. tnx FPL.
19961201 change: added note to README about the mailing list. tnx FPL.
19961201 change: documented rcpthosts wildcards. tnx RN.
19961201 change: added note to FAQ about making mailx use datemail.
19961201 change: added datemail. function requested by several people;
         approach suggested by TG.
19961201 change: added predate.
19961129 change: added QUEUE_EXTRA, QUEUE_EXTRALEN.
19961129 change: qmail-remote bounces messages with partial final lines.
19961129 change: added atomcheck() to quote crappy atoms.
19961129 change: revised atomok() to let atoms deal with more crap.
19961127 change: qmail-send adds note to deferral if flagdying. tnx TG.
19961127 change: split off maildirbounce, maildir2qmtp, and maildir2smtp
         into a separate serialmail package.
19961126 change: eliminated beta success reports from README.
19961124 change: forced res_query() return value to fit inside incoming
         buffer size. allegedly there are buggy versions of res_query()
	 that don't guarantee this.
19961122 qmail 0.93, gamma.
19961122 change: allowed empty arg list in forward.
19961121 change: qmail-smtpd now uses unknown (like qmail-qmtpd) rather
         than dying if environment variables are not set.
19961121 cleanup: reorganized helo handling in qmail-smtpd.c.
19961121 cleanup: eliminated newfield_rec.
19961121 cleanup: introduced DATE822FMT. used it in received.c.
19961121 cleanup: introduced received.c. used it in qmail-qmtpd.c,
         qmail-smtpd.c.
19961121 change: qmail-smtpd now generates a new timestamp for each
         message. tnx PJG.
19961121 cleanup: used stralloc in newfield.
19961121 cleanup: eliminated newfield_cc.
19961121 change: eliminated hfield_mort().
19961119 change: added 2-minute damper on tcpto.
19961118 change: wrote defaults for readfile controls in showctl.
19961117 change: control_readfile() now allows comments and blank lines.
         tnx LW.
19961117 change: qmail-start sets logger gid to GID_NOFILES.
19961117 bug: ipme_init() uses a fixed-length buffer for SIOCGIFCONF.
         impact: qmail-smtpd and qmail-remote will die if there are too
	 many local IP addresses. tnx MD. fix: ipme_init() now
	 dynamically allocates space, up to 200000 bytes, as long as
	 SIOCGIFCONF keeps failing. note that this is a very widespread
	 bug; it's in amd, exim, mrouted, named, nntpd, rarpd, sendmail,
	 tcpdump, timed, xntpd, and probably dozens more programs.
19961117 portability problem: on BSD 4.4 and various other systems,
         SIOCGIFCONF will truncate long lists and return success.
	 impact: on those systems, qmail-smtpd and qmail-remote will
	 miss some local IP addresses if there are too many. fix:
	 ipme_init() now checks whether there is enough space left in
	 the buffer for another ifreq, plus 64 bytes JIC. yuck.
19961117 change: ipmeprint now flushes only at end.
19961117 cleanup: introduced subfdinsmall. used it in qmail-clean.c,
         qmail-qmtpd.c.
19961117 cleanup: introduced subfdoutsmall. used it in hostname.c,
         printbreak.c, printnumd.c, printsplit.c, qmail-alias.c,
	 qmail-clean.c, qmail-getpw.c, qmail-qmtpd.c, maildir2smtp.c,
	 maildir2qmtp.c.
19961117 change: moved subfderr buf up to 256 characters.

qmail-start.c
#include "fd.h"
#include "prot.h"
#include "exit.h"
#include "fork.h"
#include "auto_uids.h"

char *(qsargs[]) = { "qmail-send", 0 };
char *(qcargs[]) = { "qmail-clean", 0 };
char *(qlargs[]) = { "qmail-lspawn", "./Mailbox", 0 };
char *(qrargs[]) = { "qmail-rspawn", 0 };
#ifdef EXTERNAL_TODO
char *(qtargs[]) = { "qmail-todo", 0};
#endif

void die() { _exit(111); }

int pi0[2];
int pi1[2];
int pi2[2];
int pi3[2];
int pi4[2];
int pi5[2];
int pi6[2];
#ifdef EXTERNAL_TODO
int pi7[2];
int pi8[2];
int pi9[2];
int pi10[2];
#endif

void close23456() {
  close(2); close(3); close(4); close(5); close(6);
#ifdef EXTERNAL_TODO
  close(7); close(8);
#endif
}

void closepipes() {
  close(pi1[0]); close(pi1[1]); close(pi2[0]); close(pi2[1]);
  close(pi3[0]); close(pi3[1]); close(pi4[0]); close(pi4[1]);
  close(pi5[0]); close(pi5[1]); close(pi6[0]); close(pi6[1]);
#ifdef EXTERNAL_TODO
  close(pi7[0]); close(pi7[1]); close(pi8[0]); close(pi8[1]);
	close(pi9[0]); close(pi9[1]); close(pi10[0]); close(pi10[1]);
#endif
}

void main(argc,argv)
int argc;
char **argv;
{
  if (chdir("/") == -1) die();
  umask(077);
  if (prot_gid(auto_gidq) == -1) die();

  if (fd_copy(2,0) == -1) die();
  if (fd_copy(3,0) == -1) die();
  if (fd_copy(4,0) == -1) die();
  if (fd_copy(5,0) == -1) die();
  if (fd_copy(6,0) == -1) die();
#ifdef EXTERNAL_TODO
  if (fd_copy(7,0) == -1) die();
  if (fd_copy(8,0) == -1) die();
#endif

  if (argv[1]) {
    qlargs[1] = argv[1];
    ++argv;
  }

  if (argv[1]) {
    if (pipe(pi0) == -1) die();
    switch(fork()) {
      case -1:
	die();
      case 0:
        if (prot_gid(auto_gidn) == -1) die();
        if (prot_uid(auto_uidl) == -1) die();
        close(pi0[1]);
        if (fd_move(0,pi0[0]) == -1) die();
        close23456();
        execvp(argv[1],argv + 1);
	die();
    }
    close(pi0[0]);
    if (fd_move(1,pi0[1]) == -1) die();
  }
 
  if (pipe(pi1) == -1) die();
  if (pipe(pi2) == -1) die();
  if (pipe(pi3) == -1) die();
  if (pipe(pi4) == -1) die();
  if (pipe(pi5) == -1) die();
  if (pipe(pi6) == -1) die();
#ifdef EXTERNAL_TODO
  if (pipe(pi7) == -1) die();
  if (pipe(pi8) == -1) die();
  if (pipe(pi9) == -1) die();
  if (pipe(pi10) == -1) die();
#endif

  switch(fork()) {
    case -1: die();
    case 0:
      if (fd_copy(0,pi1[0]) == -1) die();
      if (fd_copy(1,pi2[1]) == -1) die();
      close23456();
      closepipes();
      execvp(*qlargs,qlargs);
      die();
  }
 
  switch(fork()) {
    case -1: die();
    case 0:
      if (prot_uid(auto_uidr) == -1) die();
      if (fd_copy(0,pi3[0]) == -1) die();
      if (fd_copy(1,pi4[1]) == -1) die();
      close23456();
      closepipes();
      execvp(*qrargs,qrargs);
      die();
  }
 
  switch(fork()) {
    case -1: die();
    case 0:
      if (prot_uid(auto_uidq) == -1) die();
      if (fd_copy(0,pi5[0]) == -1) die();
      if (fd_copy(1,pi6[1]) == -1) die();
      close23456();
      closepipes();
      execvp(*qcargs,qcargs);
      die();
  }

#ifdef EXTERNAL_TODO
  switch(fork()) {
    case -1: die();
    case 0:
      if (prot_uid(auto_uids) == -1) die();
      if (fd_copy(0,pi7[0]) == -1) die();
      if (fd_copy(1,pi8[1]) == -1) die();
      close23456();
      if (fd_copy(2,pi9[1]) == -1) die();
      if (fd_copy(3,pi10[0]) == -1) die();
      closepipes();
      execvp(*qtargs,qtargs);
      die();
  }

  switch(fork()) {
    case -1: die();
    case 0:
      if (prot_uid(auto_uidq) == -1) die();
      if (fd_copy(0,pi9[0]) == -1) die();
      if (fd_copy(1,pi10[1]) == -1) die();
      close23456();
      closepipes();
      execvp(*qcargs,qcargs);
      die();
  }
#endif

  if (prot_uid(auto_uids) == -1) die();
  if (fd_copy(0,1) == -1) die();
  if (fd_copy(1,pi1[1]) == -1) die();
  if (fd_copy(2,pi2[0]) == -1) die();
  if (fd_copy(3,pi3[1]) == -1) die();
  if (fd_copy(4,pi4[0]) == -1) die();
  if (fd_copy(5,pi5[1]) == -1) die();
  if (fd_copy(6,pi6[0]) == -1) die();
#ifdef EXTERNAL_TODO
  if (fd_copy(7,pi7[1]) == -1) die();
  if (fd_copy(8,pi8[0]) == -1) die();
#endif
  closepipes();
  execvp(*qsargs,qsargs);
  die();
}
